crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2577,19 +2604,95 @@ impl<'db> TypeInferenceBuilder<'db> {
                     }
                 } else {
                     Type::LiteralString
-                }
+                };
+                Some(ty)
             }
 
             (Type::LiteralString, Type::IntLiteral(n), ast::Operator::Mult)
             | (Type::IntLiteral(n), Type::LiteralString, ast::Operator::Mult) => {
-                if n < 1 {
+                let ty = if n < 1 {
                     Type::StringLiteral(StringLiteralType::new(self.db, ""))
                 } else {
                     Type::LiteralString
+                };
+                Some(ty)
+            }
+
+            (Type::Instance(_), Type::IntLiteral(_), op) => {
+                self.infer_binary_expression_type(left_ty, KnownClass::Int.to_instance(self.db), op)
+            }
+
+            (Type::IntLiteral(_), Type::Instance(_), op) => self.infer_binary_expression_type(
+                KnownClass::Int.to_instance(self.db),
+                right_ty,
+                op,
+            ),
+
+            (Type::Instance(_), Type::Tuple(_), op) => self.infer_binary_expression_type(
+                left_ty,
+                KnownClass::Tuple.to_instance(self.db),
+                op,
+            ),
+
+            (Type::Tuple(_), Type::Instance(_), op) => self.infer_binary_expression_type(
+                KnownClass::Tuple.to_instance(self.db),
+                right_ty,
+                op,
+            ),
+
+            (Type::Instance(_), Type::StringLiteral(_), op) => {
+                self.infer_binary_expression_type(left_ty, KnownClass::Str.to_instance(self.db), op)
+            }
+
+            (Type::StringLiteral(_), Type::Instance(_), op) => self.infer_binary_expression_type(
+                KnownClass::Str.to_instance(self.db),
+                right_ty,
+                op,
+            ),
+
+            (Type::Instance(_), Type::BytesLiteral(_), op) => self.infer_binary_expression_type(
+                left_ty,
+                KnownClass::Bytes.to_instance(self.db),
+                op,
+            ),
+
+            (Type::BytesLiteral(_), Type::Instance(_), op) => self.infer_binary_expression_type(
+                KnownClass::Bytes.to_instance(self.db),
+                right_ty,
+                op,
+            ),
+
+            (Type::Instance(left_class), Type::Instance(right_class), op) => {
+                if left_class != right_class && right_ty.is_assignable_to(self.db, left_ty) {
```

Comment:
I don't think we should use `is_assignable_to` here, or even `is_subtype_of`; I think we should go straight to an `is_subclass_of` method implemented directly on `ClassType`.

Functionally it shouldn't matter currently, though it will be more efficient to bypass unnecessary layers of abstraction. But in the future with generic classes, I _think_ this decision should be based purely on the subclass relationship of the classes, without considering generic parameters at all. E.g. if class `Bar[T]` inherits class `Foo[T]` and overrides `__radd__`, then I think we should prioritize `Bar.__radd__` even if we have `Bar[X]` and `Foo[Y]`, where `Bar[X]` is not a subtype of `Foo[Y]` (for variance reasons or because X and Y have no subtype relationship), because the runtime won't be considering generic parameters (they don't even exist at runtime.)

Of course our eventual determination of which method is _supported_ will depend on the type annotations on the dunder methods, which may involve `T`, and that could influence our eventual determination of the result of the operation. But I don't think `T` should influence the decision of which dunder method to try first. Which means I think this should be based purely on the subclass relationship of two classes, without considering anything else handled in `is_assignable_to` or `is_subtype_of`.

Reply:
what if a class has `Any` in its MRO

Reply:
> what if a class has `Any`/`Unknown` in its MRO

(I should add a test for this. Meant to do so; forgot!)

Reply:
I think that the hypothetical `ClassType::is_subclass_of` should handle the special case of `Any` or `Unknown` in the MRO. (And unions too, if we support them in bases, though this might be handled via checking in multiple MROs instead.)

Reply:
Mypy and pyright both reveal `int` here:

```py
from typing import Any

class X:
    def __add__(self, other: object) -> int:
        return 42

class Y(Any):
    pass

reveal_type(X() + Y())
```

I don't actually know what the right answer is... do we need to consider `Y` as a class that "could inherit from `X` and override `__radd__`"? I think we _do_, because `Any` represents an unknown set of objects, and that unknown set of objects could include the `X` class. So a better answer than what mypy/pyright give might be `int | Any` (accounting for the fact that it _could_ be a class that inherits from `X`, but it also _could not_)?

Reply:
> Any represents an unknown set of objects, and that unknown set of objects could include the X class

I think it's even slightly more subtle than that; the relevant case here is that Any represents some unknown third class Z which inherits from X and overrides `__radd__` (since we can see that Y itself does not override `__radd__`).

I agree that `int | Any` is the right answer here. It would be `int & Any` if `X` also defined `__radd__` to return `int`.

Getting to this right answer will depend on several features, one of which is `is_subclass_of` supporting finding `Any` in the MRO, and the other is `class_member` and `inherited_class_member` supporting `Any` in the MRO.